ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/AutoCreateIcebergTagAction.java

```diff
@@ -0,0 +1,197 @@
+  private static final Logger LOG = LoggerFactory.getLogger(AutoCreateIcebergTagAction.class);
+
+  private final Table table;
+  private final TagConfig tagConfig;
+  private final LocalDateTime now;
+
+  public AutoCreateIcebergTagAction(Table table, LocalDateTime now) {
+    this.table = table;
+    this.tagConfig = TagConfig.fromTableProperties(table.properties());
+    this.now = now;
+  }
+
+  public void execute() {
+    if (!tagConfig.isAutoCreateTag()) {
+      return;
+    }
+    if (tagExist()) {
+      LOG.debug("{} find expect tag, skip", table.name());
+      return;
+    }
+    boolean success = createTag();
+    LOG.info("{} tag creation {}", table.name(), success ? "succeed" : "skipped");
+  }
+
+  private boolean tagExist() {
+    if (TagTriggerPeriod.DAILY.propertyName().equals(tagConfig.getTriggerPeriod())) {
+      return findTagOfToday() != null;
+    } else {
+      throw new IllegalArgumentException(
+          "unsupported trigger period " + tagConfig.getTriggerPeriod());
+    }
+  }
+
+  private String findTagOfToday() {
+    String name = generateTagName();
+    return table.refs().entrySet().stream()
+        .filter(entry -> entry.getValue().isTag())
+        .map(Map.Entry::getKey)
+        .filter(name::equals)
+        .findFirst()
+        .orElse(null);
+  }
+
+  private boolean createTag() {
+    Snapshot snapshot = findSnapshot(table, getTagTriggerTime());
+    if (snapshot == null) {
+      LOG.info("{} no snapshot found at {}", this.table.name(), getTagTriggerTime());
+      return false;
+    }
+    String newTagName = generateTagName();
+    table.manageSnapshots().createTag(newTagName, snapshot.snapshotId()).commit();
+    LOG.info(
+        "{} create tag {} on snapshot {} {}",
+        this.table.name(),
+        newTagName,
+        snapshot.snapshotId(),
+        snapshot.timestampMillis());
+    return true;
+  }
+
+  private String generateTagName() {
+    if (TagTriggerPeriod.DAILY.propertyName().equals(tagConfig.getTriggerPeriod())) {
+      String tagFormat = tagConfig.getTagFormat();
+      return now.minusDays(1).format(DateTimeFormatter.ofPattern(tagFormat));
+    } else {
+      throw new IllegalArgumentException(
+          "unsupported trigger period " + tagConfig.getTriggerPeriod());
+    }
+  }
+
+  private long getTagTriggerTime() {
+    if (TagTriggerPeriod.DAILY.propertyName().equals(tagConfig.getTriggerPeriod())) {
+      int triggerOffsetMinutes = tagConfig.getTriggerOffsetMinutes();
+      LocalTime localTime = LocalTime.ofSecondOfDay(triggerOffsetMinutes * 60L);
+      LocalDateTime localDateTime = LocalDateTime.of(now.toLocalDate(), localTime);
+      return localDateTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();
+    } else {
+      throw new IllegalArgumentException(
+          "unsupported trigger period " + tagConfig.getTriggerPeriod());
+    }
+  }
+
+  private static Snapshot findSnapshot(Table table, long tagTriggerTime) {
```

Comment:
If I start the amoro at 15:00 and snapshot keep it for 12 hours, then I don't get the snapshot closest to 0:00, I think is not allowed here!

Reply:
Yes, this is caused by the snapshots expiring service, maybe we should fix this in another PR and keep these snapshots from expiring.

BTW, in this case, do you think we should give up creating the tag or tagging on the snapshot of 03:00?

Reply:
> Yes, this is caused by the snapshots expiring service, maybe we should fix this in another PR and keep these snapshots from expiring.
> 
> BTW, in this case, do you think we should give up creating the tag or tagging on the snapshot of 03:00?

I think this is unacceptable

Reply:
Maybe we should introduce a config like `tag.auto-create.max-delay`, when the time of the earliest snapshot exceeds this delay, give up creating the tag.

Reply:
> Maybe we should introduce a config like `tag.auto-create.max-delay`, when the time of the earliest snapshot exceeds this delay, give up creating the tag.

+1

Reply:
I introduced a table property `tag.auto-create.trigger.max-delay.minutes`.